File: anvil-api/src/main/java/rocks/milspecsg/anvil/api/repository/Repository.java
/*
 *   Anvil - MilSpecSG
 *   Copyright (C) 2020 Cableguy20
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rocks.milspecsg.anvil.api.repository;

import rocks.milspecsg.anvil.api.model.ObjectWithId;
import rocks.milspecsg.anvil.api.storageservice.StorageService;

import java.time.Instant;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public interface Repository<
    TKey,
    T extends ObjectWithId<TKey>,
    TDataStore>
    extends StorageService<TKey, T, TDataStore> {

    /**
     * @return The time of creation of this document as an {@link Instant}
     */
    CompletableFuture<Optional<Instant>> getCreatedUtc(TKey id);
}

File: core/src/main/scala/org/anvilpowered/anvil/core/db/Repository.scala
/*
 *   Anvil - AnvilPowered.org
 *   Copyright (C) 2019-2026 Contributors
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.core.db

import cats.data.ReaderT

import java.util.UUID
import cats.data.OptionT
import cats.effect.Async

trait Repository[E <: DomainEntity] {
  def findById[F[_]: Async](uuid: UUID): OptionT[F, E]

  def exists[F[_]: Async](uuid: UUID): OptionT[F, Boolean]

  def countAll[F[_]: Async]: F[Long]

  def deleteById[F[_]: Async](id: UUID): F[Boolean]
}
